source/position.go

```diff
@@ -0,0 +1,63 @@
+// Copyright © 2024 Meroxa, Inc. and Miquido
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package source
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/conduitio/conduit-commons/opencdc"
+)
+
+// Position represents position of a document in an index.
+type Position struct {
+	mu             sync.Mutex
+	IndexPositions map[string]int `json:"indexPositions"`
```

Comment:
* The map IndexPositions is a map of the index and the position till the search API fetches the record from that index. 
* The struct Position stores the list of such indexes and the last fetched position. 
*  Elasticseach internally stores an integer `_seq_no` and returns a sorted list based on this integer. So when we restart the connector, the map tells us to use the int position of an index as offset to fetch values after that.

Reply:
Here's my understanding of this works, at least in theory, please let me know if something's wrong. So, IIUC from the ES docs, a sequence number represents the number of _operations_ performed in an index. Every update and deletion will count towards that.

On the other hand, when we create a search request in `internal/elasticsearch/v8/search.go`, we use the sequence as the `From` field, that is used to skip a certain number of _documents_ from the search results. Since the number of operations and the number of documents do not have to match, we might be skipping some documents. E.g.:

1. create doc 1 (_seq_no = 0)
2. update doc 1 (_seq_no = 1)
3. update doc 1 again (_seq_no = 2)
4. create doc 2 (_seq_no = 3)
5. stop connector
6. create doc 3 (_seq_no = 4)
7. start connector

IIUC, when the connector starts, it will use the last sequence number, which is 3 and it will skip 3 documents. That means it will skip doc 3, which is not that we want.

Reply:
Currently, this is how it works. Just think of it as calling a paginated API. Let's say we currently have only one index to fetch data from, let's call it users.
1. Connector Starts 
             1.1 IndexPositions map is {"users": 0}
             1.2 we are polling /users/_search?size=5&from=0
2. We create a new doc1 in index users
            2.1 Now  /users/_search?size=5&from=0 will return doc1
            2.2 Update the IndexPositions map as {"users": 1}
3. Now it is polling /users/_search?size=5&from=1 
4. We create a new doc2 in index users
            2.1 Now  /users/_search?size=5&from=1 will return doc2
            2.2 Update the IndexPositions map as {"users": 2}  
5. STOP connector
6. Create doc3, doc4, doc5
7. START connector again 
           7.1 IndexPositions map is  now {"users": 2}
           7.2 calling /users/_search?size=5&from=2 will return doc3, doc4, doc5. 

So IndexPositions is a map of index and a counter(internal to our source connector) that tells how many records have been read and from(skip) where we should start reading ahead.

Reply:
Yup, I understand that, and I believe it will work for created documents just fine. I'm just not sure (or maybe I don't fully understand) how it works for updated and deleted documents as in the example I mentioned.

Reply:
Ok, I understand your concern. Currently, we can fetch only the new records here since Elasticsearch doesn't provide any way to capture change data which is useful for our source implementation.

Reply:
Yup, I understand we can't get updates and deletes in the Elasticsearch source. However, I'm concerned about what happens in the connector if a user updates or deletes documents. It looks like we might loose some created documents too?

Reply:
I have tested it multiple times,  tried covering all cases. It works fine.

Reply:
Really cool, thanks!